sources/api/apiserver/src/server/ephemeral_storage.rs

```diff
@@ -0,0 +1,380 @@
+    for entry in entries {
+        let entry = entry.context(error::DiscoverEphemeralSnafu {})?;
+        filenames.push(entry.path().into_os_string().to_string_lossy().to_string());
+    }
+    Ok(filenames)
+}
+
+/// allowed_bind_dirs returns a set of the directories that can be bound to ephemeral storage, which
+/// varies based on the variant
+pub fn allowed_bind_dirs(variant: &str) -> HashSet<&'static str> {
+    let mut allowed = HashSet::from(["/var/lib/containerd", "/var/lib/host-containerd"]);
+    if variant.contains("k8s") {
+        allowed.insert("/var/lib/kubelet");
+        allowed.insert("/var/log/pods");
+    }
+    if variant.contains("ecs") {
+        allowed.insert("/var/lib/docker");
+        allowed.insert("/var/log/ecs");
+    }
+    allowed
+}
+
+/// scans the raid array to identify if it has been created already
+fn mdadm_scan() -> Result<Vec<u8>> {
+    let output = Command::new(MDADM)
+        .args([OsStr::new("--detail"), OsStr::new("--scan")])
+        .output()
+        .context(error::ExecutionFailureSnafu { command: MDADM })?;
+    ensure!(output.status.success(), error::ScanArrayFailureSnafu {});
+    Ok(output.stdout)
+}
+
+/// is_formatted returns true if the array is already formatted with the specified filesystem
+pub fn is_formatted<S: AsRef<OsStr>>(device: S, format: &Filesystem) -> Result<bool> {
+    let mut fmt_arg = OsString::from("TYPE=");
+    fmt_arg.push(OsString::from(format.to_string()));
+
+    let blkid = Command::new(BLKID)
+        .args([
+            OsStr::new("--match-token"),
+            fmt_arg.as_ref(),
+            device.as_ref(),
+        ])
+        .status()
+        .context(error::DetermineFormatFailureSnafu {})?;
+
+    Ok(blkid.success())
+}
+
+/// formats the specified device with the given filesystem format
+pub fn format_device<S: AsRef<OsStr>>(device: S, format: &Filesystem) -> Result<()> {
+    let binary = match format {
+        Filesystem::Xfs => MKFSXFS,
+        Filesystem::Ext4 => MKFSEXT4,
+    };
+
+    let mut mkfs = Command::new(binary);
+    mkfs.arg(device.as_ref());
+    // labeled, XFS has a max of 12 characters, EXT4 allows 16
+    mkfs.arg("-L");
+    mkfs.arg(EPHEMERAL);
+
+    let mkfs = mkfs
+        .status()
+        .context(error::ExecutionFailureSnafu { command: binary })?;
+
+    ensure!(mkfs.success(), error::FormatFilesystemFailureSnafu {});
+    Ok(())
+}
+
+pub mod error {
+    use snafu::Snafu;
+
+    #[derive(Debug, Snafu)]
+    #[snafu(visibility(pub(super)))]
+    pub enum Error {
+        #[snafu(display("Failed to execute '{:?}': {}", command, source))]
+        ExecutionFailure {
+            command: &'static str,
+            source: std::io::Error,
+        },
+
+        #[snafu(display("Failed to discover ephemeral disk"))]
+        DiscoverEphemeral { source: std::io::Error },
+
+        #[snafu(display("Failed to mount array (non-zero exit)"))]
+        MountArrayFailure {},
```

Comment:
These "non-zero exit" errors could use more context, including the error message from the command:
```suggestion
        #[snafu(display("Failed to mount {} to {}: {}", what, where, output))]
        MountArrayFailure { what: PathBuf, where: PathBuf, output: std::process::Output },
```